contracts/Pool.sol

```diff
@@ -1,31 +1,292 @@
+        internal
+    {
+        uint64 blockNumber = uint64(block.number);
+        Sender storage sender = senders[msg.sender];
+        // Iterating over receivers, see `ReceiverWeights` for details
+        address receiverAddr = ReceiverWeightsImpl.ADDR_ROOT;
+        while (true) {
+            uint32 weight = 0;
+            (receiverAddr, weight) = sender.receiverWeights.nextWeight(
+                receiverAddr
+            );
+            if (weight == 0) break;
+            Receiver storage receiver = receivers[receiverAddr];
+            // The receiver was never used, initialize it
+            if (
+                amtPerWeightPerBlockDelta > 0 &&
+                receiver.lastCollectedCycle == 0
+            ) receiver.lastCollectedCycle = blockNumber / cycleBlocks;
+            int256 delta = int256(weight) * amtPerWeightPerBlockDelta;
+            setDeltaRange(receiver.amtDeltas, blockNumber, blockEnd, delta);
+        }
     }
 
-    function withdraw(uint256 value) public {
-        withdrawTo(value, msg.sender);
+    /// @notice Sets delta of a single receiver in a given block range
+    /// @param amtDeltas The deltas of the per-cycle receiving rate
+    /// @param blockStart The block number from which the delta takes effect
+    /// @param blockEnd The block number from which the delta stops taking effect
+    /// @param delta Change of the per-block receiving rate
+    function setDeltaRange(
+        mapping(uint64 => int256) storage amtDeltas,
+        uint64 blockStart,
+        uint64 blockEnd,
+        int256 delta
+    ) internal {
+        setSingleDelta(amtDeltas, blockStart, delta);
+        setSingleDelta(amtDeltas, blockEnd, -delta);
     }
 
-    function withdrawTo(uint256 value, address payable receiver) public {
-        uint256 balance = senders[msg.sender].balance;
-        require(value <= balance, "Not enough funds in account");
-        balance -= value;
-        senders[msg.sender].balance = balance;
-        receiver.transfer(value);
+    /// @notice Sets delta of a single receiver on a given block number
+    /// @param amtDeltas The deltas of the per-cycle receiving rate
+    /// @param blockNumber The block number from which the delta takes effect
+    /// @param delta Change of the per-block receiving rate
+    function setSingleDelta(
+        mapping(uint64 => int256) storage amtDeltas,
+        uint64 blockNumber,
+        int256 delta
+    ) internal {
+        // In order to set a delta on a specific block it must be introduced in two cycles.
+        // The cycle delta is split proportionally based on how much the first cycle is affected.
+        // The second cycle has the rest of the delta applied, so the update is fully completed.
+        int256 cycle2Blocks = int256(blockNumber) % cycleBlocks;
+        int256 cycle1Blocks = cycleBlocks - cycle2Blocks;
+        uint64 cycle1 = blockNumber / cycleBlocks + 1;
+        uint64 cycle2 = cycle1 + 1;
```

Comment:
These variable names could be more descriptive
```suggestion
        uint64 currentCycle = blockNumber / cycleBlocks + 1;
        uint64 nextCycle = currentCycle + 1;
```

Reply:
I would avoid using `current`, it suggests that it has something to do with the current block. At this level of genericness we really don't know much about these cycles except that they go one after another. It's a shortened version of `firstCycle` and `secondCycle`.